tips/tip-0001/tip-0001.md

```diff
@@ -0,0 +1,67 @@
+
+Let $g$ and $h$ be uniformly random scalars. These scalars will be used as initials to randomize the `rp` and `fd` columns when zero-knowledge is an objective. When zero-knowledge is not necessary, it cannot hurt to set $g = 1$ and $h=0$. Also note that since $h$ does not need to be the formal derivative of $g$, and if this is the case the name "formal derivative" for the column is not entirely accurate. However, it may help at first reading to imagine these scalars as polynomials satisfying $h(X) = g'(X)$.
+
+The running product starts with $g$ initially and accumulates a factor $X - \mathsf{ramp}$ in every pair of rows where $\mathsf{ramp} \neq \mathsf{ramp}^\star$. This evolution corresponds to one transition constraint: $(\mathsf{ramp}^\star - \mathsf{ramp}) \cdot (\mathsf{rp}^\star - \mathsf{rp} \cdot (X - \mathsf{ramp})) + (1 -(\mathsf{ramp}^\star -\mathsf{ramp}) \cdot \mathsf{di}) \cdot (\mathsf{rp}^\star - \mathsf{rp})$.
+
+The associated terminal value is $T_{\mathsf{rp}}$ and the terminal constraint is stipulates that the factor $X - \mathsf{ramp}$ from the last row is accumulated as well: $\mathsf{rp} \cdot (X - \mathsf{ramp}) - T_{\mathsf{rp}}$. Alternatively, if the last row is a padding row and should not be accumulated into the running product, then the last value in the column is already the terminal value: $\mathsf{rp} - T_{\mathsf{rp}}$.
+
+The formal derivative contains the "formal derivative" of the running product with respect to $X$ (ignoring the masking factors $g$ and $h$). The formal derivative is initially $h$. The transition constraint applies the product rule of differentiation conditioned upon the difference in `ramp` being nonzero; in other words, if $\mathsf{ramp} = \mathsf{ramp}^\star$ then the same value persists; but if $\mathsf{ramp} \neq \mathsf{ramp}^\star$ then $\mathsf{fd}$ mapped as
+
+$$ \mathsf{fd} \mapsto \mathsf{fd}^ \star = (X - \mathsf{ramp}) \cdot \mathsf{fd} + \mathsf{rp}   . $$
+
+This update rule is called the *product rule of differentiation* because when ignoring $g(X)$ and $h(X)$ or when setting $h(X) = g'(X)$ and assuming $\mathsf{ramp}^\star \neq \mathsf{ramp}$, then
+
+$$ \frac{\mathsf{d}  \mathsf{rp}^\star}{\mathsf{d}   X} = \frac{\mathsf{d}  (X - \mathsf{ramp}) \cdot \mathsf{rp}}{\mathsf{d}   X} $$
+$$ = (X - \mathsf{ramp}) \cdot \frac{\mathsf{d}   \mathsf{rp}}{\mathsf{d}   X} + \frac{\mathsf{d}  ( X - \mathsf{ramp})}{\mathsf{d}   X} \cdot \mathsf{rp} $$
+$$ = (X - \mathsf{ramp}) \cdot \mathsf{fd} +\mathsf{rp} \enspace . $$
+
+The transition constraint for $\mathsf{fd}$ is $(\mathsf{ramp}^\star - \mathsf{ramp}) \cdot (\mathsf{fd}^\star - \mathsf{rp} - (X - \mathsf{ramp}) \cdot \mathsf{fd}) + (1 -(\mathsf{ramp}^\star -\mathsf{ramp}) \cdot \mathsf{di}) \cdot (\mathsf{fd}^\star - \mathsf{fd})$.
```

Comment:
This is a personal preference and I'm open to counter arguments: I'm adding $\mathsf{di}$ as a factor to the first summand here to
1. make the two summands symmetric, and
1. keep the leading term's constant unchanged.

The degree of the constraint polynomial does not increase, but implementing the `extend` function becomes (arguably marginally) simpler.

Reply:
I think the degree of the constraint polynomials do increase. That was exactly my motivation for dropping the redundant factor.

Reply:
You are right, it does increase the degree. My sense for symmetry is bummed out, but I withdraw the suggestion.